src/utils/sealed_file.rs

```diff
@@ -0,0 +1,67 @@
+//! A file whose fd cannot be written by other processes
+//!
+//! This mechanism is useful for giving clients access to large amounts of
+//! information such as keymaps without them being able to write to the handle.
+
+use nix::{
+    fcntl::{FcntlArg, SealFlag},
+    sys::memfd::MemFdCreateFlag,
+};
+use std::{
+    ffi::CString,
+    fs::File,
+    io::{Seek, Write},
+    os::unix::prelude::{AsRawFd, FromRawFd, RawFd},
+};
+
+#[derive(Debug)]
+pub(crate) struct SealedFile {
+    file: File,
+    size: usize,
+}
+
+impl SealedFile {
+    pub fn new(name: &str, contents: &str) -> Result<Self, std::io::Error> {
+        let name = CString::new(name).expect("File name should not contain interior null bytes");
+        let contents =
+            CString::new(contents).expect("Sealed file contents should not contain interior null bytes");
```

Comment:
Not sure if I like the fact that it now can panic behind callers back. In rust string, interior null bytes are perfectly valid, although rare, so maybe it does not matter.

I would either accept the CStr, or propagate the NullError to the caller.
SealError perhaps?
```rs
#[derive(thiserror::Error, Debug)]
pub enum SealError{
       Io(std::io::Error),
       InvalidName(std::ffi::NulError),
       InvalidContent(std::ffi::NulError),
}
```

Reply:
yea i was wondering what to do, i think that's good